### ISSUE TITLE
App rejected because of LegacyNotificationTrigger

### ISSUE THREAD

**Issue by sachinB94** [OneSignal/OneSignal-iOS-SDK#345]
App store rejection with the following message:

```
Your app uses or references the following non-public APIs:

- "UNLegacyNotificationTrigger, UIConcreteLocalNotification"
```

Must be due to this line here:
https://github.com/OneSignal/OneSignal-iOS-SDK/blob/2e5d1762cc49b4e4fecaf6298b3f04df70b5d640/iOS_SDK/OneSignalSDK/Source/UNUserNotificationCenter%2BOneSignal.m#L244

**Comment by Nightsd01** [comment]
Hi @sachinB94 how did you add the OneSignal iOS SDK to your project?

Can you try replacing the method `callLegacyAppDelegateSelector:...` with the following code:

```objc
+ (void)callLegacyAppDeletegateSelector:(UNNotification *)notification
                            isTextReply:(BOOL)isTextReply
                       actionIdentifier:(NSString*)actionIdentifier
                               userText:(NSString*)userText
                fromPresentNotification:(BOOL)fromPresentNotification
                  withCompletionHandler:(void(^)())completionHandler {
    [OneSignal onesignal_Log:ONE_S_LL_VERBOSE message:@"callLegacyAppDeletegateSelector:withCompletionHandler: Fired!"];
    
    UIApplication *sharedApp = [UIApplication sharedApplication];
    
    // trigger is nil when UIApplication.presentLocalNotificationNow: is used.
    //  However it will be UNLegacyNotificationTrigger when UIApplication.scheduleLocalNotification: is used
    BOOL isLegacyLocalNotif = !notification.request.trigger || [notification.request.trigger isKindOfClass:NSClassFromString([NSString stringWithFormat:@"UNLegacy%@", @"NotificationTrigger"])];
    BOOL isCustomAction = actionIdentifier && ![@"com.apple.UNNotificationDefaultActionIdentifier" isEqualToString:actionIdentifier];
    BOOL isRemote = [notification.request.trigger isKindOfClass:NSClassFromString(@"UNPushNotificationTrigger")];
    
    if (isLegacyLocalNotif) {
        UILocalNotification *localNotif = [NSClassFromString([NSString stringWithFormat:@"UIConcrete%@", @"LocalNotification"]) alloc];
        localNotif.alertBody = notification.request.content.body;
        localNotif.alertTitle = notification.request.content.title;
        localNotif.applicationIconBadgeNumber = [notification.request.content.badge integerValue];
        NSString* soundName = [notification.request.content.sound valueForKey:@"_toneFileName"];
        if (!soundName)
            soundName = @"UILocalNotificationDefaultSoundName";
        localNotif.soundName = soundName;
        localNotif.alertLaunchImage = notification.request.content.launchImageName;
        localNotif.userInfo = notification.request.content.userInfo;
        localNotif.category = notification.request.content.categoryIdentifier;
        localNotif.hasAction = true; // Defaults to true, UNLocalNotification doesn't seem to have a flag for this.
        localNotif.fireDate = notification.date;
        localNotif.timeZone = [notification.request.trigger valueForKey:@"_timeZone"];
        localNotif.repeatInterval = (NSCalendarUnit)[notification.request.trigger valueForKey:@"_repeatInterval"];
        localNotif.repeatCalendar = [notification.request.trigger valueForKey:@"_repeatCalendar"];
        // localNotif.region =
        // localNotif.regionTriggersOnce =
        
        if (isTextReply &&
            [sharedApp.delegate respondsToSelector:@selector(application:handleActionWithIdentifier:forLocalNotification:withResponseInfo:completionHandler:)]) {
            NSDictionary* dict = @{UIUserNotificationActionResponseTypedTextKey: userText};
            [sharedApp.delegate application:sharedApp handleActionWithIdentifier:actionIdentifier forLocalNotification:localNotif withResponseInfo:dict completionHandler:^() {
                completionHandler();
            }];
        }
        else if (isCustomAction &&
                 [sharedApp.delegate respondsToSelector:@selector(application:handleActionWithIdentifier:forLocalNotification:completionHandler:)])
            [sharedApp.delegate application:sharedApp handleActionWithIdentifier:actionIdentifier forLocalNotification:localNotif completionHandler:^() {
                completionHandler();
            }];
        else if ([sharedApp.delegate respondsToSelector:@selector(application:didReceiveLocalNotification:)]) {
            [sharedApp.delegate application:sharedApp didReceiveLocalNotification:localNotif];
            completionHandler();
        }
        else
            completionHandler();
    }
    else if (isRemote) {
        NSDictionary* remoteUserInfo = notification.request.content.userInfo;
        
        if (isTextReply &&
            [sharedApp.delegate respondsToSelector:@selector(application:handleActionWithIdentifier:forRemoteNotification:withResponseInfo:completionHandler:)]) {
            NSDictionary* responseInfo = @{UIUserNotificationActionResponseTypedTextKey: userText};
            [sharedApp.delegate application:sharedApp handleActionWithIdentifier:actionIdentifier forRemoteNotification:remoteUserInfo withResponseInfo:responseInfo completionHandler:^() {
                completionHandler();
            }];
        }
        else if (isCustomAction &&
                 [sharedApp.delegate respondsToSelector:@selector(application:handleActionWithIdentifier:forRemoteNotification:completionHandler:)])
            [sharedApp.delegate application:sharedApp handleActionWithIdentifier:actionIdentifier forRemoteNotification:remoteUserInfo completionHandler:^() {
                completionHandler();
            }];
        // Always trigger selector for open events and for non-content-available receive events.
        //  content-available seems to be an odd expection to iOS 10's fallback rules for legacy selectors.
        else if ([sharedApp.delegate respondsToSelector:@selector(application:didReceiveRemoteNotification:fetchCompletionHandler:)] &&
                 (!fromPresentNotification ||
                 ![[notification.request.trigger valueForKey:@"_isContentAvailable"] boolValue])) {
            // NOTE: Should always be true as our AppDelegate swizzling should be there unless something else unswizzled it.
            [sharedApp.delegate application:sharedApp didReceiveRemoteNotification:remoteUserInfo fetchCompletionHandler:^(UIBackgroundFetchResult result) {
                // Call iOS 10's compleationHandler from iOS 9's completion handler.
                completionHandler();
            }];
        }
        else
            completionHandler();
    }
    else
        completionHandler();
}
```

**Comment by sachinB94** [comment]
Hey @Nightsd01 
```
[NSString stringWithFormat:@"UNLegacy%@", @"NotificationTrigger"]
```
I read it somewhere that this thing doesn't really works. They still find out.

The previous versions of our app were also being accepted fine, with OneSignal SDK included.
But, they just started pointing out these issue when I'm submitting the build now.

Any ideas, why are they doing it now?

**Comment by Nightsd01** [comment]
@sachinB94 We have a suggestion to avoid this issue. 

In your AppDelegate, can you set your AppDelegate to be a `UNUserNotificationCenterDelegate` and implement the following two methods (you can leave them as empty stubs):

***Objective-C***
```objc
-(void)userNotificationCenter:(UNUserNotificationCenter *)center willPresentNotification:(UNNotification *)notification withCompletionHandler:(void (^)(UNNotificationPresentationOptions))completionHandler {
    
}

-(void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)(void))completionHandler {
    
}
```

***Swift***
```swift
func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any], fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
    
}
    
func application(_ application: UIApplication, handleActionWithIdentifier identifier: String?, for notification: UILocalNotification, withResponseInfo responseInfo: [AnyHashable : Any], completionHandler: @escaping () -> Void) {
    
}
```

We are using this API in order to maintain backwards compatibility. Essentially, there are some AppDelegate methods (such as `application:handleActionWithIdentifier:forLocalNotification:withResponseInfo:completionHandler:`) which stop getting called if you use the UNNotificationCenter. 

In order to maintain backwards compatibility for many developers who still rely on these methods, we manually call these selectors if the developer has implemented them.

However, if you have a `UNUserNotificationCenterDelegate` and you implement those two methods, those private-API codepaths will never get executed and we suspect Apple's system to detect private API usage will not catch it.

Please let me know if this does not resolve your issue. Thanks!

**Comment by wojtekdmyszewicz** [comment]
@Nightsd01 I'm trying your solution, but I'm afraid it will be rejected again. Because when I run `strings - -a -arch armv7 "YourApp.app/YourApp" | grep UNLegacyNotificationTrigger` on the binary it still returns two instances.

**Comment by wojtekdmyszewicz** [comment]
They keep complaining 

Specifically, this app's [OneSignalUNUserNotificationCenter callLegacyAppDeletegateSelector:isTextReply:actionIdentifier:userText:fromPresentNotification:withCompletionHandler:] method calls the non-public API 'UIConcreteLocalNotification' as well as using several non-public selectors of the non-public class. It would be appropriate to completely remove all non-public API use from this app before resubmitting for review.

**Comment by Nightsd01** [comment]
@sachinB94 @wojtekdmyszewicz

This issue has now been addressed in #350 and has been included in the new release, 3.8.0. Thank you for taking the time to report it and I hope it didn’t cause you too much trouble!

**Comment by sachinB94** [comment]
@Nightsd01 Great, thanks man. I did push the app for review after removing LegacyNotificationTrigger from the SDK. It was accepted. But, I'll pull 3.8.0 now.
